### PR TITLE
add a cache buster for the browser

### PR DESCRIPTION
**Pull request by mszulik**

### CODE REVIEW

File: src/Controller/ApiController.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * Updates models with information provided after a video was transcoded by the Transmorpher media server.
+     *
+     * @param array $videoTranscodingResult
+     * @return void
+     */
+    protected function handleVideoTranscodingNotification(array $videoTranscodingResult): void
+    {
+        // If we have no upload for the provided token, the derivatives were purged on the media server, and we should update the hash of the latest upload.
```
Comment (gael-connan-cybex) [inline src/Controller/ApiController.php:65]:
```suggestion
        // If we receive a transcoding result, but we don't know the upload token, that means that the server has purged and recreated the derivatives. We need to update the media hash on our side for cache busting.
```

File: src/Enums/ServerNotification.php [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Transmorpher\Enums;
+
+enum ServerNotification: string
```
Comment (gael-connan-cybex) [inline src/Enums/ServerNotification.php:5]: like, seen from here, isnt it a client notification?

File: src/Image.php [hunk 3]
```diff
@@ -70,25 +69,36 @@ public function getDerivative(array $transformations = []): string
...
-     * Sends the request to reserve an upload slot to the Transmorpher media server API.
-     *
-     * @return Response
+     * @return string
      */
-    protected function sendReserveUploadSlotRequest(): Response
+    public function getThumbnailUrl(): string
     {
-        return $this->configureApiRequest()->post(TransmorpherApi::S2S->getUrl('image/reserveUploadSlot'), ['identifier' => $this->getIdentifier()]);
+        return $this->getUrl(['height' => 300]);
```
Comment (gael-connan-cybex) [inline src/Image.php:102]: should we make thumbnail height configurable

File: src/Media.php [hunk 4]
```diff
@@ -114,7 +117,7 @@ public function upload($fileHandle, string $fileName = null): array
                     );
             }
 
-            $clientResponse = $this->getClientResponseFromResponse($response);
+            $clientResponse = $this->getClientResponseFromServerResponse($response);
```
Comment (gael-connan-cybex) [inline src/Media.php:120]: i dont get it. how many requests are happening?

File: src/Models/TransmorpherMedia.php [hunk 5]
```diff
@@ -64,4 +73,18 @@ public function getMedia(): Media
     {
         return $this->type->getMediaClass()::for($this->Transmorphable, $this->media_name);
     }
+
+    public function isAvailable(): Attribute
+    {
+        return Attribute::make(
+            get: fn(): bool => $this->is_ready && $this->public_path
```
Comment (gael-connan-cybex) [inline src/Models/TransmorpherMedia.php:80]: wasnt that renamed to full size url somewhere

File: src/Models/TransmorpherUpload.php [hunk 6]
```diff
@@ -80,4 +80,11 @@ public function handleStateUpdate(array $response, int $httpCode = null): array
 
         return $response;
     }
+
+    public function isLatest(): Attribute
```
Comment (gael-connan-cybex) [inline src/Models/TransmorpherUpload.php:84]: if this isnt a protocol, and the rows are overwritten, how could there be an older version. and if its just an outdated version in ram, wouldnt something like is() just compare the id, which might not change?

File: src/Media.php (hunk 4 above)
Comment (mszulik) [reply src/Media.php:120]: What do you mean? It's 1 request to reserve an upload slot and afterwards we are uploading the image.

File: src/Image.php (hunk 3 above)
Comment (mszulik) [reply src/Image.php:102]: good idea

File: src/Media.php [hunk 7]
```diff
@@ -233,10 +229,8 @@ public function setVersion(int $versionNumber): array
         $upload = $this->transmorpherMedia->TransmorpherUploads()->create(['state' => UploadState::INITIALIZING, 'message' => 'Sending request to restore version.']);
 
         try {
-            $response = $this->configureApiRequest()->patch(TransmorpherApi::S2S->getUrl(sprintf('media/%s/version/%s', $this->getIdentifier(), $versionNumber)), [
-                'callback_url' => sprintf('%s/%s', config('transmorpher.api.callback_base_url'), config('transmorpher.api.callback_route')),
-            ]);
-            $clientResponse = $this->getClientResponseFromResponse($response);
+            $serverResponse = $this->configureApiRequest()->patch(TransmorpherApi::S2S->getUrl(sprintf('media/%s/version/%s', $this->getIdentifier(), $versionNumber)));
+            $clientResponse = $this->getClientResponseFromServerResponse($serverResponse);
```
[the remark this answers is not in the document]
Comment (mszulik) [reply src/Media.php:233]: I just renamed the variable and function.

If you mean the client response stuff - we are checking if the server response is 'success', if not, we are modifying the response to return a client friendly and understandable response to the frontend.

File: src/Models/TransmorpherMedia.php (hunk 5 above)
Comment (mszulik) [reply src/Models/TransmorpherMedia.php:80]: We are passing a full size URL to the frontend now, but there is still a field`public_path` in the TransmorpherMedia table

File: src/Models/TransmorpherUpload.php (hunk 6 above)
Comment (mszulik) [reply src/Models/TransmorpherUpload.php:84]: The rows are only overwritten when stuff happens to one specific upload, there is still one row per upload / action.

File: src/Image.php [hunk 8]
```diff
@@ -99,6 +99,6 @@ public function getUrl(array $transformations = []): string
      */
     public function getThumbnailUrl(): string
     {
-        return $this->getUrl(['height' => 300]);
+        return $this->getUrl(['height' => $this->getThumbnailHeight()]);
```
Comment (gael-connan-cybex) [inline src/Image.php:102]: you can configure anything as transformation, but only height is applied?

Comment (mszulik) [reply src/Image.php:102]: Currently only height is configurable. I added a comment to clarify that.

**Review by gael-connan-cybex — APPROVED** [review]
tested again, upload, restore, deletion

**Pull request merged**